File: C/2015/day01.c
#include "readFileToString.h"
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define DATA_FILE "day01.dat"

int day01() {
    const char *filename = DATA_FILE;
    char *fileContents = readFileToString(filename);

    if (fileContents) {
        // printf("%s\n", fileContents);
        int floor = 0;
        int position = 0;
        bool first_time = true;
        char *move = fileContents;
        for (unsigned long i = 0; i < strlen(fileContents); i++, move++) {
            position++;
            if (*move == ')') {
                floor--;
            } else {
                floor++;
            }
            if ((floor == -1) && (first_time == true)) {
                first_time = false;
                printf("2015 Day 01 Part 2: %d\n", position);
            }
            // printf("%c", *move);
        }

        printf("2015 Day 01 Part 1: %d\n", floor);
        free(fileContents);
    }

    return 0;
}

File: python/2015/day01.py
# python day01.py ../../data/2015/day01.dat

import sys


def parse_input(input_data):
    """
    Parse the input data from the Advent of Code problem.
    Edit this function to fit the day's input format.
    """
    return input_data.strip().split("\n")


def part1(data):
    """
    Solution for Part 1.
    """
    floor = 0
    for move in data[0]:
        if move == ")":
            floor -= 1
        else:
            floor += 1

    return floor


def part2(data):
    """
    Solution for Part 2.
    """
    floor = 0
    position = 0
    for move in data[0]:
        position += 1
        if move == ")":
            floor -= 1
        else:
            floor += 1
        if floor == -1:
            return position

    return None


if __name__ == "__main__":
    # Read input file or use test string
    if len(sys.argv) > 1:
        with open(sys.argv[1], "r") as f:
            input_data = f.read()
    else:
        # Example test input, replace with actual test data
        input_data = """\
))(((((
"""

    data = parse_input(input_data)
    print("Day 01 Part 1:", part1(data))
    print("Day 01 Part 2:", part2(data))
